src/librustc_middle/ty/instance.rs

```diff
@@ -502,6 +483,62 @@ impl<'tcx> Instance<'tcx> {
     }
 }
 
+fn polymorphize<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    substs: SubstsRef<'tcx>,
+) -> SubstsRef<'tcx> {
+    debug!("polymorphize({:?}, {:?})", def_id, substs);
+    let unused = tcx.unused_generic_params(def_id);
+    debug!("polymorphize: unused={:?}", unused);
+
+    struct PolymorphizationFolder<'tcx> {
+        tcx: TyCtxt<'tcx>,
+    };
+
+    impl ty::TypeFolder<'tcx> for PolymorphizationFolder<'tcx> {
+        fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+            self.tcx
+        }
+
+        fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+            debug!("fold_ty: ty={:?}", ty);
+            match ty.kind {
+                ty::Closure(def_id, substs) => {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_closure(def_id, polymorphized_substs)
+                }
+                ty::Generator(def_id, substs, movability) => {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_generator(def_id, polymorphized_substs, movability)
+                }
+                _ => ty.super_fold_with(self),
```

Comment:
This is not that related to this PR itself, but why aren't we doing the same for `FnDef` here?

Reply:
At least for the original mangling issue, I wasn't sure that it could happen with a `FnDef`.

Reply:
this should allow us to optimize the following I think
```rust
fn unused<T>() -> u64 {
    42
}

fn test<T>() {
    std::iter::repeat(unused::<T>).take(10);
}

pub fn main() {
    test::<String>();
    test::<Vec<String>>();
}
```

Reply:
This probably also has the mangling issue :thinking: not really sure about how mangling works for `FnDef` though

Reply:
Fixed!